### DISCUSSION TITLE
Book metadata format

### DISCUSSION THREAD

**Discussion by Jdiesel87** [advplyr/audiobookshelf#59] [Q&A]
I have bounced between a few different audiobook server products (Plex, Emby, Booksonic, audioserve) and they all seem to treat metadata differently. Will Audiobookshelf use ID3v2 tags to store metadata moving forward? I only ask as it can be quite a chore to move from one method to the next.

My library is currently setup for booksonic which includes a cover.jpg, desc.txt, and reader.txt in the root directly which is a bit limiting so I welcome the option for storing more information for each book.

Can you speak to how the currently available fields correspond to the ID3v2 tags?

For example 
Author=Artist
Narrarator=Composer

**Comment by advplyr** [comment]
The only place where file metadata is used is in the downloader when creating `m4b` files. In that case it will embed the cover art and set the title, artist (author) and date (publish year). This should be easy to expand on and include other metadata, you can [see in the code](https://github.com/advplyr/audiobookshelf/blob/master/server/utils/ffmpegHelpers.js#L41) the function is a bit bare at the moment.

I think you point out something important, the scanner should definitely be checking all the metadata fields to populate the audiobook details (the scanner does use the track # metadata tag already to try to get the correct order of the tracks).
It is important to note as well that AudioBookshelf does not add/remove/modify anything in your audiobooks directory currently.

As someone who has a lot of experience with the other media servers, I would really appreciate your input on this. If you could outline your ideal scenario that would be a great place to start.

**Comment by Jdiesel87** [comment]
I will put something together on an example use case and try to spread the word to some other folks who might have input too.

I think the right way to do it is likely to store the metadata in an external nfo. I have been burnt in the past where another application has overwritten my previously curated ID3 tags. Being able to import and export to a nfo makes life easier and allows one to share a library of media between different applications without compatibility issues.

Is/will scraping be within the scope of this project?

**Comment by Jdiesel87** [comment]
I spent a bit of time looking over my library and thinking about my ideal situation for metadata management. I realize that there is no single correct way to do things but this is what I currently have. About 90% of my library is in m4b format. I still have a few books in multitrack mp3 format that I will eventually combine into a single m4b when time permits. The majority of my books originate from Audible but I am not the one personally ripping them so existing tags are not always complete.

Artist and Album Artist: Author
Title and track name: Book title
Composer: Narrator
Media Kind: audiobook
Genre: actual genre tags 
Description: Book synopsis
Release date: release date of audiobook
Cover: Cover art

I haven't been using it but possible Series and Series Part could be useful to some people. 

I also have a desc.txt for each book containing a summary of the book. I find the description tag is almost never populated when I grab a new audiobook and must be added manually. In addition to booksonic using the desc.txt, I believe some m4b formatting tools like m4b-tool can use the desc.txt to bake in the description automatically when combing multiple mp3's into a m4b.

My preference would be to scan the tags during import and output a single nfo to be stored in the book directory with all the relevant metadata needed in Audiobookshelf. This way the original files are not modified.

**Reply by advplyr** [reply]
I need to put some thought into this, but question first:

Is desc.txt a normal naming convention for other software? Or is booksonic just picking it up because it is a text file?

**Reply by Jdiesel87** [reply]
I've only come across it it in booksonic

**Reply by advplyr** [reply]
So I wrote a little script to generate an `nfo` file from the audiobook data. However, from my understanding NFO files are for the user and not intended to be interpreted again by software.
For that reason I structured it to be human readable.

![image](https://user-images.githubusercontent.com/67830747/135186584-0711e4b4-9a82-4b42-9b51-1d861902be49.png)

I think the important part of handling metadata is going to be pulling it from the audio files and using it in AudioBookshelf. Then users can optionally have an `nfo` file written like above and when downloading audiobooks as `m4b` it will be included.

**Reply by advplyr** [reply]
The challenge with sourcing metadata from the audio files is going to be matching metadata fields with what they actually are.

I'm looking through my audiobooks and there isn't a standard being followed. Sometimes the author, not the narrator, is stored in the `conductor` tag.

It looks like the description is sometimes stored in the `comments` tag.

We may need to set the standard if there is none already.

**Reply by Jdiesel87** [reply]
Honestly that is half the problem and also why this project has so much potential, there is no standard and no leader in the audiobook space to set a standard. If you ask 10 people how they organize their books you will get 5 different answers.

I'm guessing that Plex probably has the biggest market share for self hosted audiobook software right now. Making it easy for those users to easily migrate to Audiobookshelf will likely be a major win.

It may just take setting a standard, sticking too it, and becoming the leader in this space.

**Comment by Jdiesel87** [comment]
There is some good information in this document

https://github.com/seanap/Plex-Audiobook-Guide

**Reply by Jdiesel87** [reply]
An example of Audible tags and mapping to their respective ID3 tags from the link above

https://github.com/seanap/Plex-Audiobook-Guide#tags-that-are-being-set

**Reply by advplyr** [reply]
What do you think the Year tag should represent? In their example they are using the audiobook release year basically.
In AudioBookshelf I meant for publish year to be the books first publish year.

Another question, do you think AudioBookshelf should support "Release Year" and "Publish Year"?

**Reply by Jdiesel87** [reply]
I personally have been tagging my books with the book release year. I don't have much value in audiobook release year but there may be some use cases for it that fall outside of mine own. I know there can be multiple releases of the same book but usually the narrator can be used to distinguish between the different versions.

**Reply by advplyr** [reply]
Of the ~100 audiobooks I scanned the majority of them that have a year in the metadata are using the audiobook release year.
The scanner could just always store the year in metadata as the publish year, assuming the publish year isn't in the title folder name.

**Reply by advplyr** [reply]
Just did a huge revamp of the scanner to use the metadata in files to fill out the book details.

Audiobooks that have embedded cover art get the cover art extracted to `/metadata/books/{BOOK ID}/cover.*`, and that cover is used in AudioBookshelf now.

Next the scanner will look for a `desc.txt` and set the description if that is there.

After all the data is set from the directory structure, then it checks the audio file metadata and populates details from that. The file metadata won't override anything already set to this point, so it is used only if the detail field is empty.

These are the mappings it checks for so far:

```javascript
    const MetadataMapArray = [
      {
        tag: 'tagComposer',
        key: 'narrarator'
      },
      {
        tag: 'tagDescription',
        key: 'description'
      },
      {
        tag: 'tagPublisher',
        key: 'publisher'
      },
      {
        tag: 'tagDate',
        key: 'publishYear'
      },
      {
        tag: 'tagSubtitle',
        key: 'subtitle'
      },
      {
        tag: 'tagArtist',
        key: 'author'
      }
    ]
```

I cleared my audiobooks and ran a fresh scan and saw a huge improvement already. A lot of covers set and extra details set that weren't getting pulled before.

I pushed this under `v1.2.9`, but since I would consider this  "minor" release I'm going to update the docs and release this as `v1.3.0` tomorrow.

**Reply by Jdiesel87** [reply]
Just scanned my library and it picked up all my desc.txt perfectly.

I don't think it is bringing in the tags correctly though. I looked at a few books that have the date, composer, and artist tags populated but the details fields in Audiobookshelf remain blank after a scan.

The log file doesn't show much. Is there a way to increase the verbose level for debugging purposes?

**Reply by advplyr** [reply]
It only scrapes the metadata for new audiobooks. Have you tried with removing the audiobook and scanning again?

**Reply by Jdiesel87** [reply]
That did the trick.

**Reply by Jdiesel87** [reply]
One thing to note, removing and rescanning will cause you to lose your playback progress. My environment is for testing but any users who have been using the software for a while now might not be too keen on losing that history.

**Reply by advplyr** [reply]
Agree, I did 2 things to try to remedy this for existing users. 
1. running scan will check if the last scan was before version 1.3.1 and re-scan metadata, and 
2. added a "Re-Scan" button on the audiobook details modal.

The re-scan button is going to check the embedded cover art and the metadata on audio files again

I updated the [website docs](https://www.audiobookshelf.org/docs#metadata) with the metadata mapping info, but it is pretty weak. At some point these docs are going to need to be organized better.

**Comment by advplyr** [comment]
Another update related to metadata in [v1.3.4](https://github.com/advplyr/audiobookshelf/releases/tag/v1.3.4)

You can now opt to store cover art in your audiobook directory.

And cover art from URLs are downloaded and saved, so http:// urls won't be set on audiobooks going forward. This is important because those URLs can change or get taken down, and it is also better to not be making all those requests outside your server to get cover art.

**Comment by Jdiesel87** [comment]
I don't know if you are still using the OpenLibrary API to scrape for info but what are your thoughts on grabbing the GoodReads ID if the API returns it? If it can't be automated with a scan even having a field in the details sections would be helpful for now. I see it being presented something like this where the ID is a link to the relevant GoodReads page.
![tempsnip](https://user-images.githubusercontent.com/14824476/138740249-e20eb521-ac7f-4053-b282-77aa03397ccd.png)

**Reply by Jdiesel87** [reply]
Or potentially just a icon or badge instead of the ID number

**Reply by advplyr** [reply]
I removed the match tab that pulled details from open library, it just wasn't reliable enough. The only thing getting pulled from outside sources right now is the covers. Still on the lookout for a more reliable way to fetch book data, and still thinking it might become a separate project. If we do end up pulling in goodreads id then I like the idea of a little icon that links to the page.
I saw Dune last night, thought it was great, never read the series though. Considering reading the series now, but my reading list is a mile long.

**Reply by Jdiesel87** [reply]
I'm about 8 hours into the first book, out of about 20. I wanted to finish it before seeing the new movie but I'm getting impatient and might just cave in and watch it.

Yeah it is really a pain. Maybe I will have a peak at some of the Calibre scraper plugins to see what they are doing. Those scraper plugins often are in defiance with the ToS of the providers though, probably why they aren't included in the core software. Like you had mentioned before, there is really a need for an opensource book database with an API but that is a huge undertaking not to mention the hosting costs associated with it.

**Reply by jarrodCoombes** [reply]
| I removed the match tab that pulled details from open library, it just wasn't reliable enough

Can you expand on that? As in the info was not accurate, or it just didn't pull anything down? 

Would scraping from Wikipedia not be an option (is that even possible)?

**Reply by advplyr** [reply]
Reliable meaning that using a title and an author you can get back data in a usable format the vast majority of time. With the free open source book databases I found that grabbing the book cover was the only practical use, so that is what audiobookshelf uses to lookup cover images.
Web scraping requires continual maintenance as websites change their structure over time. There is also the issue of IP rate limiting that most sites put in place. Lastly, there could be potential intellectual property issues that I wouldn't want to come across, I'm not sure about that though.

**Reply by jarrodCoombes** [reply]
Can openlibrary get us the ISBN? If so, that might be worth adding in somewhere, and maybe allow for an ISBN list export. Then you can use Calibre to get all the meta data based on the ISBN (or some other tool).

**Reply by Jdiesel87** [reply]
Openlibrary does return the ISBN in the results but the problem is still linking the correct text valve used in the search to the right book. If we search by ISBN the results are much more accurate.

Maybe what could be done is include a way to manually enter the ISBN for a book. I feel like once the correct ISBN is linked to the book and metadata search results will be much more reliable. This is kind of what Emby does with Movies and TV series but letting you identify media by IMDB ID, TVDB ID, The MovieDB ID rather than Title and/or year.
![Capture](https://user-images.githubusercontent.com/14824476/139082696-c9925dc0-0627-4b6f-a075-cf7585a46a05.PNG)

**Reply by jarrodCoombes** [reply]
Oh yes, I think the ISBN is key here, there definitely should be a field for it in the meta data that a user can then edit.

> but the problem is still linking the correct text valve used in the search to the right book

I am not sure I follow, are you saying that the actual result from OpenLibrary can be questionable, or are you saying the returned values from open library are questionable?

The cover results are pretty good IMO, there were very few that wrong when I added all my books, less than 5%, and while that's probably not great for retrieving all the meta data, it's probably good enough to auto populate the ISBN number.

What about when Audiobookshelf grabs the cover, it also grabs the ISBN and throws into a user reviewable field in the meta data editor. Then a user can go into each book, check the ISBN and click a button that says "grab meta data based on this ISBN" and then manually reviews the result and either cancels or accepts it. 

It's still a manual process, but it's way better than having to copy and past things for thousands of books, and it does mean the results don't have to be perfect, just good enough that most times the "grab meta data" button gets the correct info.

**Reply by advplyr** [reply]
With movies and series there is 1 id for IMDB and TMDB. With ISBN a book can have dozens of ISBN's of 13 or 10 digits that come up with unpredictable results in my testing. The original Book class had an ISBN field.
The book descriptions that come out of open library are also a mess about half of the time on my library.

I'm open to adding the ISBN field back, but where do you suggest we fetch metadata from with that ISBN?

**Reply by Jdiesel87** [reply]
My thought was to continue using openlibrary with the hope that searching by ISBN would help improve the results but if the book description coming out of openlibrary are a mess it isn't really a proper solution. I don't really have an alternative suggestion.

**Reply by Jdiesel87** [reply]
Maybe the Google Books API?

https://developers.google.com/books/docs/v1/using

https://www.googleapis.com/books/v1/volumes?q=return+of+the+king+inauthor:tolkien

https://www.googleapis.com/books/v1/volumes?q=isbn=9780007488346

**Reply by advplyr** [reply]
I've used google books API it is the best one available right now I think, but I thought all of the requests required an API key.
It looks like this could be an option actually since the volume search is free. I did a quick search in the docs and didn't find what their rate limits are. I suppose we could put it in experimental features for a while and see how it works out.

I had completely dismissed google books api because I thought it was pay only, nice find

**Reply by advplyr** [reply]
Google books turned out to be a success so far. I released book matching as experimental in `v1.5.6` and started a new discussion #157 

It is much faster then open library and so far I'm getting better results, especially for description.

Let me know what you think and if you are getting accurate results.

**Comment by wtanksleyjr** [comment]
I posted a discussion about this, but research indicates that this thread might be more appropriate.

The NFO creation seems to have been removed in the latest build. Is there any plan for a replacement? Audiobookself is a wonderful metadata editor, especially with its "Match" support. But without some way to export my edits, I can't feed them back to the rest of my scripts, and since the mobile app doesn't work for me, I have to export the metadata to use it.

**Reply by advplyr** [reply]
I don't plan on adding it back in, the new abmetadata file is a much better solution. If you explain your use-case in more detail maybe we can find another way for you to do what you need to do.

**Reply by wtanksleyjr** [reply]
Thank you!

I like the abmetadata file, but it's missing one thing: a button to CHOOSE to save it to the book folder just for that one book. That button was removed when NFO format was removed. I don't miss NFO, but I do miss that button.

The use case is that most books are automatically added to the database based on extracted metadata and folder names, so they're generally "good enough." But sometimes I comb through books, for example by looking at everything a given author wrote, and notice a book's out of its series (super common, the providers almost never get that right). So I fix it, probably by clicking "Match", but also by editing directly. Anyhow, I want my scripts to know about the change ... so I used to click the "Save Metadata to Folder" button. That means the created file is _golden_.

This isn't as good if the abmetadata is ALWAYS in the book's folder (as with the option). That means I have to decide whether I edited the metadata or not, and it turns out it's REALLY hard to get all the little corner cases inside a script.

OTOH, it would be (for my use case) just as good if the abmetadata file had a flag that meant something like "manually inspected and approved." I'd be fine with that.

**Reply by advplyr** [reply]
What about using the ASIN or ISBN fields to determine whether it was correctly matched?

**Comment by wtanksleyjr** [comment]
Not bad, and if I want to edit by hand I can just make something up. (That
will happen with podcast books.)

**Comment by openaudible** [comment]
It looks like you're using an extension of the FFMPEG "metadata" output as the meta data format. There are a few gotchas with that as a format. 

First, it isn't simple to parse, as the chapters require special handing (so you can't simply generate key=value pairs for everything) like a normal property file. 

Secondly, properly escaping the characters is a little funky if you want to use ffmpeg's standard. 
https://ffmpeg.org/ffmpeg-all.html#toc-Quoting-and-escaping
Things break easily when you try to use both quotes and escape characters. 

But except for the rare case of leading or trailing spaces, you can conform to the ffmpeg standard by just escaping =;#\n and \ characters with a leading \. 

But I didn't see any escaping going on in a first glance at your code... so not sure.

Lastly, the key=value format gets a little ugly when you're dealing with lots of data and possibly multi-line data. 

I personally json is a better format.. its pretty easy to read as a human and most of the time a computer is going to read it and render it.

**Reply by advplyr** [reply]
Agree JSON is easier to work with in code. The reason I started with a file similar to ffmpegs is because I thought it would be easier for the user to read it and update it as necessary. I am not using the exact specifications of ffmetadata, I was just basing a similar metadata format off of that, so I'm not sure a lot of that applies.

I did have to work with ffmetadata when embedding metadata in with ffmpeg and passing in a text file with a list of audio files for transcoding. https://github.com/advplyr/audiobookshelf/blob/master/server/utils/ffmpegHelpers.js#L7

I am calling the metadata format abmetadata and the generator and parser is here: https://github.com/advplyr/audiobookshelf/blob/master/server/utils/abmetadataGenerator.js
This is not meant to be used in ffmpeg. 
Any metadata that is multiple lines is put in its own section like how chapters are done in ffmetadata. Currently only description could have multiple lines.

Here is an example abmetadata file:
```
;ABMETADATA2
#audiobookshelf v2.0.7

media=book
title=Wizards First Rule
subtitle=
authors=Terry Goodkind
narrators=Sam Tsoutsouvas
publishedYear=2008
publisher=Brilliance Audio
isbn=
asin=B002V0QK4C
language=
genres=Audiobook Fantasy
series=Sword of Truth
explicit=N

[DESCRIPTION]
The masterpiece that started Terry Goodkind's New York Times bestselling epic Sword of Truth In the aftermath of the brutal murder of his father, a mysterious woman, Kahlan Amnell, appears in Richard Cypher's forest sanctuary seeking help...and more.

[CHAPTER]
start=0
end=6004.6675
title=Terry Goodkind - SOT Bk01 - Wizards First Rule 01
[CHAPTER]
start=6004.6675
end=12000.946
title=Terry Goodkind - SOT Bk01 - Wizards First Rule 02
[CHAPTER]
start=12000.946
end=17997.250625
title=Terry Goodkind - SOT Bk01 - Wizards First Rule 03
```

This is harder to parse in scripts, but easier to read and update for users. I'm not set on using this, I could be convinced to just use JSON or YAML.

**Comment by wtanksleyjr** [comment]
Question: right now I'm pregenerating OPF when I move folders into
Audiobookshelf's folder, it seems to be pretty reliable. What would happen
if I instead created one of these metadata files instead? Would my
generated file be looked at first, and loaded without further
modifications? That would actually simplify my scripts.

-Wm

**Reply by advplyr** [reply]
If the abmetadata file is saved as `metadata.abs`  inside your library item folder then it will be used as the source of truth for metadata.

I wouldn't say that file is finalized yet, it is possible we move to a different format of metadata. Maybe JSON is better, maybe YAML, not sure. There isn't really a standard to go off of so we must create one.

**Comment by wtanksleyjr** [comment]
I'll start doing that, then ... I suspect it'll make it easier to have just
a single source of truth.

Yeah, I hear you about the "no standard format." It's amazing to me that
it's taken this long to come up with an audiobook server. The various past
efforts have been just other products rebadged (although Booksonic-Air
deserves a salute for doing as well as it does). I'll keep an eye here for
format changes, certainly.

If you asked me, and you shouldn't, I'd pick yaml for a file describing a
single book, and JSON for a file describing a lot of books. I abhor XML.

-Wm

**Comment by Dev-Nitin0** [comment]
यू नhee
